mars/oscar/backends/ray/communication.py

```diff
@@ -0,0 +1,278 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import concurrent.futures as futures
+import itertools
+from abc import ABC
+from collections import namedtuple
+from enum import Enum
+from typing import Any, Callable, Coroutine, Dict, Type
+from urllib.parse import urlparse
+
+from ..communication.base import Channel, ChannelType, Server, Client
+from ..communication.core import register_client, register_server
+from ..communication.errors import ChannelClosed
+from ....utils import implements, classproperty
+from ....utils import lazy_import
+
+ray = lazy_import("ray")
+
+DEFAULT_DUMMY_RAY_ADDRESS = 'ray://0'
+
+
+class MessageType(Enum):
+    SEND = 0
+    RECV = 1
+
+
+ChannelID = namedtuple("ChannelID", ["local_address", "dest_address", "channel_index"])
+
+
+class RayChannelBase(Channel, ABC):
+    """
+    Channel for communications between ray processes.
+    """
+    __slots__ = '_channel_index', '_channel_id', '_in_queue', '_closed'
+
+    name = 'ray'
+    _channel_index_gen = itertools.count()
+
+    def __init__(self,
+                 local_address: str = None,
+                 dest_address: str = None,
+                 channel_index: int = None,
+                 channel_id: ChannelID = None,
+                 compression=None):
+        super().__init__(local_address=local_address,
+                         dest_address=dest_address,
+                         compression=compression)
+        self._channel_index = channel_index or next(self._channel_index_gen)
+        self._channel_id = channel_id or ChannelID(local_address, dest_address, self._channel_index)
+        self._in_queue = asyncio.Queue()
+        self._closed = asyncio.Event()
+
+    @property
+    def channel_id(self) -> ChannelID:
+        return self._channel_id
+
+    @property
+    @implements(Channel.type)
+    def type(self) -> ChannelType:
+        return ChannelType.ray
+
+    @implements(Channel.close)
+    async def close(self):
+        self._closed.set()
+
+    @property
+    @implements(Channel.closed)
+    def closed(self) -> bool:
+        return self._closed.is_set()
+
+
+class RayClientChannel(RayChannelBase):
+    """A channel from ray driver/actor to ray actor. Use ray call reply for client channel recv.
+    """
+    __slots__ = '_peer_actor',
+
+    def __init__(self,
+                 dest_address: str = None,
+                 channel_index: int = None,
+                 channel_id: ChannelID = None,
+                 compression=None):
+        super().__init__(None, dest_address, channel_index, channel_id, compression)
+        # ray actor should be created with the address as the name.
+        self._peer_actor: 'ray.actor.ActorHandle' = ray.get_actor(dest_address) if dest_address else None
```

Comment:
The `RayClientChannel` should asserts that the peer actor is not None. So, just `self._peer_actor = ray.get_actor(dest_address)`?